src/main/java/com/zendesk/maxwell/producer/MaxwellPubsubProducer.java

```diff
@@ -0,0 +1,225 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+class PubsubCallback implements ApiFutureCallback<String> {
+  public static final Logger LOGGER = LoggerFactory.getLogger(PubsubCallback.class);
+
+  private final AbstractAsyncProducer.CallbackCompleter cc;
+  private final Position position;
+  private final String json;
+  private final Timer timer;
+  private MaxwellContext context;
+
+  private Counter succeededMessageCount;
+  private Counter failedMessageCount;
+  private Meter succeededMessageMeter;
+  private Meter failedMessageMeter;
+
+  public PubsubCallback(AbstractAsyncProducer.CallbackCompleter cc, Position position, String json,
+                        Timer timer, Counter producedMessageCount, Counter failedMessageCount, Meter producedMessageMeter,
+                        Meter failedMessageMeter, MaxwellContext context) {
+    this.cc = cc;
+    this.position = position;
+    this.json = json;
+    this.timer = timer;
+    this.succeededMessageCount = producedMessageCount;
+    this.failedMessageCount = failedMessageCount;
+    this.succeededMessageMeter = producedMessageMeter;
+    this.failedMessageMeter = failedMessageMeter;
+    this.context = context;
+  }
+
+  @Override
+  public void onSuccess(String messageId) {
+    this.succeededMessageCount.inc();
+    this.succeededMessageMeter.mark();
+
+    if ( LOGGER.isDebugEnabled() ) {
+      LOGGER.debug("->  " + this.json);
+      LOGGER.debug("    " + position);
+      LOGGER.debug("");
+    }
+
+    cc.markCompleted();
+    timer.update(cc.timeToSendMS(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void onFailure(Throwable t) {
+    this.failedMessageCount.inc();
+    this.failedMessageMeter.mark();
+
+    LOGGER.error(t.getClass().getSimpleName() + " @ " + position);
+    LOGGER.error(t.getLocalizedMessage());
+
+    if ( !this.context.getConfig().ignoreProducerError ) {
+      this.context.terminate(new RuntimeException(t));
+      return;
+    }
+
+    cc.markCompleted();
+    timer.update(cc.timeToSendMS(), TimeUnit.MILLISECONDS);
+  }
+
+}
+
+
+public class MaxwellPubsubProducer extends AbstractProducer {
+  public static final Logger LOGGER = LoggerFactory.getLogger(MaxwellPubsubProducer.class);
+
+  private final ArrayBlockingQueue<RowMap> queue;
+  private final MaxwellPubsubProducerWorker worker;
+
+  public MaxwellPubsubProducer(MaxwellContext context, Properties pubsubProperties, String pubsubProjectId, String pubsubTopic) throws IOException {
+    super(context);
+    this.queue = new ArrayBlockingQueue<>(100);
+    this.worker = new MaxwellPubsubProducerWorker(context, pubsubProperties, pubsubProjectId, pubsubTopic, this.queue);
+    this.worker.setPubsub();
+    this.worker.setDDLPubsub();
+    Thread thread = new Thread(this.worker, "maxwell-pubsub-worker");
+    thread.setDaemon(true);
+    thread.start();
+  }
+
+  @Override
+  public void push(RowMap r) throws Exception {
+    this.queue.put(r);
+  }
```

Comment:
The main reason we use an intermediate queue in the kafka producer is that kafka's `send()` method can block the current thread (performing compression I believe) even though it's otherwise async.

If pubsub producer is properly async and doesn't block the sending thread at all, you could probably make this producer a bit simpler by avoiding the queue (and the worker thread), like the Kinesis producer.

Reply:
Yup, either is fine - we can always simplify later if someone does the testing to figure out whether it affects throughput 👍